src/Storages/MergeTree/MergeTreeRangeReader.cpp

```diff
@@ -921,6 +922,39 @@ bool MergeTreeRangeReader::isCurrentRangeFinished() const
     return prev_reader ? prev_reader->isCurrentRangeFinished() : stream.isFinished();
 }
 
+
+/// When executing ExpressionActions on an empty block, it is not possible to determine the number of rows
```

Comment:
does empty block mean that all rows were filtered out? if yes than we probably just need to stop processing it?

Reply:
Actually empty block in this case means that is has all nullptrs for all columns. This is the case when a column (or columns) that we want to read is not present in the part and we need to compute default value for it. We request MergeTreeReader to read some no-zero number of rows but MergeTreeReader will put nullptr in Columns vector at that position for that column and later it is computed in https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeRangeReader.cpp#L990
If all columns are not present in the part then the block doen't have the information of number of rows, so evaluateMissingDefaults will create columns with 0 rows. Dummy columns helps to pass the actual number of rows and create columns with default values of the appropriate size.

Reply:
ok, but why this didn't create problems before this patch?

Reply:
One existing physical column is always added to the required_columns.
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeBlockReadUtils.cpp#L121-L124
It is enough if we have just one DAG